core/src/banking_stage.rs

```diff
@@ -1077,6 +1015,300 @@ impl BankingStage {
         }
     }
 
+    /// Filter out packets that fail to sanitize, or are no longer valid (could be
+    /// too old, a duplicate of something already processed). Doing this in batches to avoid
+    /// checking bank's blockhash and status cache per transaction which could be bad for performance.
+    /// Added valid and sanitized packets to forwarding queue.
+    pub fn filter_and_forward_with_account_limits(
+        bank: &Arc<Bank>,
+        buffered_packet_batches: &mut UnprocessedPacketBatches,
+        forward_buffer: &mut ForwardPacketBatchesByAccounts,
+        batch_size: usize,
+        slot_metrics_tracker: &mut LeaderSlotMetricsTracker,
+        banking_stage_stats: &BankingStageStats,
+    ) -> FilterForwardingResults {
+        let mut total_forwardable_tracer_packets: usize = 0;
+        let mut total_tracer_packets_in_buffer: usize = 0;
+        let mut total_forwardable_packets: usize = 0;
+        let mut dropped_tx_before_forwarding_count: usize = 0;
+
+        let mut original_priority_queue = Self::swap_priority_queue(buffered_packet_batches);
```

Comment:
I don't think this API belongs in banking_stage because it breaks the encapsulation of the inner fields of `UnprocessedPacketBatches`. I.e. we're modifying the inner `packet_priority_queue` without modifying the `message_hash_to_transaction`, which breaks consistency. This is a precedent I would like to avoid.


Ideally all the struct fields inside `UnprocessedPacketBatches` are private, and the only mutable modifications to those fields are via exposed methods on the struct that keep both the inner `message_hash_to_transaction` and `packet_priority_queue` consistent. Implementing a `drain_desc` on `UnprocessedPacketBatches` would be appropriate here.

Reply:
Yeah, agree. This swap/modify/put-back isn't ideal. The same mechanism used here also used in `consume_buffered_packets()`. Would be cleaner to refactor it in separate PR?